File: app/src/test/java/com/kirchhoff/example/githubclient/presenters/CommitsPresenterTest.java
package com.kirchhoff.example.githubclient.presenters;

import com.kirchhoff.example.githubclient.Constants;
import com.kirchhoff.example.githubclient.repository.FakeGitHubRepository;
import com.kirchhoff.example.githubclient.repository.GitHubDataSource;
import com.kirchhoff.example.githubclient.ui.commit.CommitsContract;
import com.kirchhoff.example.githubclient.ui.commit.CommitsPresenter;
import com.kirchhoff.example.githubclient.utils.schedulers.BaseSchedulerProvider;
import com.kirchhoff.example.githubclient.utils.schedulers.ImmediateSchedulerProvider;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.MockitoAnnotations;

import static junit.framework.Assert.assertNotNull;

/**
 * @author Kirchhoff-
 */

@RunWith(JUnit4.class)
public class CommitsPresenterTest {

    @Mock
    CommitsContract.View view;

    private CommitsPresenter presenter;

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.initMocks(this);

        GitHubDataSource repository = new FakeGitHubRepository();
        BaseSchedulerProvider provider = new ImmediateSchedulerProvider();

        presenter = new CommitsPresenter(repository, view, provider);
    }

    @Test
    public void testCreated() throws Exception {
        assertNotNull(presenter);
        Mockito.verifyNoMoreInteractions(view);
    }

    @Test
    public void testCommitsError() throws Exception {
        Constants.DATA_TEST_ENUM = Constants.DataTestEnum.ERROR;
        presenter.loadCommits(Constants.REPOSITORY);

        Mockito.verify(view).showLoading();
        Mockito.verify(view).hideLoading();
        Mockito.verify(view).showError();

        Mockito.verifyNoMoreInteractions(view);
    }

    @Test
    public void testCommitsEmpty() throws Exception {
        Constants.DATA_TEST_ENUM = Constants.DataTestEnum.EMPTY;
        presenter.loadCommits(Constants.REPOSITORY);

        Mockito.verify(view).showLoading();
        Mockito.verify(view).hideLoading();
        Mockito.verify(view).showEmptyView();

        Mockito.verifyNoMoreInteractions(view);
    }

    @Test
    public void testCommitsSuccess() throws Exception {
        Constants.DATA_TEST_ENUM = Constants.DataTestEnum.DATA;
        presenter.loadCommits(Constants.REPOSITORY);

        Mockito.verify(view).showLoading();
        Mockito.verify(view).hideLoading();
        Mockito.verify(view).openCommits(Constants.emulateCommitsList());

        Mockito.verifyNoMoreInteractions(view);
    }
}

File: app/src/test/java/com/kirchhoff/example/githubclient/presenters/CommitsPresenterTest.kt
package com.kirchhoff.example.githubclient.presenters

import com.kirchhoff.example.githubclient.Constants
import com.kirchhoff.example.githubclient.Injection
import com.kirchhoff.example.githubclient.ui.commit.CommitsContract
import com.kirchhoff.example.githubclient.ui.commit.CommitsPresenter
import org.junit.Assert.assertNotNull
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import org.mockito.Mock
import org.mockito.Mockito.verify
import org.mockito.Mockito.verifyNoMoreInteractions
import org.mockito.MockitoAnnotations

/**
 * @author Kirchhoff-
 */
@RunWith(JUnit4::class)
class CommitsPresenterTest {

    @Mock
    private lateinit var view: CommitsContract.View

    private lateinit var presenter: CommitsPresenter

    @Before
    @Throws(Exception::class)
    fun init() {
        MockitoAnnotations.initMocks(this)

        presenter = CommitsPresenter(Injection.provideGitHubRepository(), view,
                Injection.provideSchedulerProvider())
    }

    @Test
    @Throws(Exception::class)
    fun testCreated() {
        assertNotNull(presenter)

        verifyNoMoreInteractions(view)
    }

    @Test
    @Throws(Exception::class)
    fun testCommitsError() {
        Constants.DATA_TEST_ENUM = Constants.DataTestEnum.ERROR
        presenter.loadCommits(Constants.REPOSITORY)

        verify(view).showLoading()
        verify(view).hideLoading()
        verify(view).showError()

        verifyNoMoreInteractions(view)
    }

    @Test
    @Throws(Exception::class)
    fun testEmptyCommits() {
        Constants.DATA_TEST_ENUM = Constants.DataTestEnum.EMPTY
        presenter.loadCommits(Constants.REPOSITORY)

        verify(view).showLoading()
        verify(view).hideLoading()
        verify(view).showEmptyView()

        verifyNoMoreInteractions(view)
    }

    @Test
    @Throws(Exception::class)
    fun testCommitsSuccess() {
        Constants.DATA_TEST_ENUM = Constants.DataTestEnum.DATA
        presenter.loadCommits(Constants.REPOSITORY)

        verify(view).showLoading()
        verify(view).hideLoading()
        verify(view).showCommits(Constants.emulateCommitsList())

        verifyNoMoreInteractions(view)
    }
}